cpp/src/lance/io/filter.h

```diff
@@ -0,0 +1,59 @@
+//  Copyright 2022 Lance Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+#pragma once
+
+#include <arrow/compute/exec/expression.h>
+#include <arrow/result.h>
+
+#include <memory>
+#include <tuple>
+
+#include "lance/format/schema.h"
+
+namespace lance::io {
+
+class FileReader;
+
+/// Filter.
+class Filter {
```

Comment:
Just curious how similar this is to Parquet? Do you have a clearer idea on where to draw the boundary between what's in the format vs compute layer at this point?

Reply:
The current plan is that, format / storage api covers single data source project + filter (including index based filter as well).  So the pylance API and pyarrow API can pass enough filters to accelerate I/Os.
Do not plan to cover things like computation, join, catelog and etc.

This is different to parquet. For parquet, there is no row-level filtering (i.e., here we are using the mask created from the filter to fetch other columns). In parquet, all the projected column/row groups are read into memory before applying filters.